### ISSUE TITLE
Bug: prefer-const does not recognize array destructuring

### ISSUE THREAD

**Issue by Slapbox** [eslint/eslint#17088]
### Environment

Node version: 16.18.0
npm version: 8.19.2
Local ESLint version: 8.38.0
Global ESLint version: 8.38.0
Operating System: Ubuntu 22.10

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
[[Occurs with recommended config for core rules]]
```
</details>

```js
  let preposition;
  [, preposition = '', myMutableString] = 		myMutableString.match(prepositionRegex);
```

### What did you expect to happen?

It should not flag `preposition` for `prefer-const` because it is reassigned.

### What actually happened?

A reassigned variable is flagged for `prefer-const`. If you make it a const, it will be recognized as reassigned and flagged for that rule instead.

### Link to Minimal Reproducible Example

No time for this currently, only reporting

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by fasttime** [comment]
Thanks @Slapbox. In your example there is only one assignment, so I think the rule is correct to report a problem. Note that the string on the right side of `preposition = ''` is a default expression, and it is only evaluated when the destructured value is `undefined`. This can be better seen with a function call:

```js
let myMutableString;
let preposition;
[, preposition = getEmptyString(), myMutableString] = ['foo bar', 'foo', 'bar'];

function getEmptyString() {
    console.log('getEmptyString called');
    return '';
}
```

In the code above, it's easy to see that `getEmptyString()` is never called.

Also, I don't understand this sentence:

> If you make it a const, it will be recognized as reassigned and flagged for that rule instead.

In your example, replacing `let`, with `const` produces unparseable code. On the other hand, if `preposition` is assigned in a `const` declaration, the rule reports no error as expected (**[DEMO](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci1jb25zdDogXCJlcnJvclwiICovXG5cbmNvbnN0IFssIHByZXBvc2l0aW9uID0gJycsIG15TXV0YWJsZVN0cmluZ10gPSBteU11dGFibGVTdHJpbmcubWF0Y2gocHJlcG9zaXRpb25SZWdleCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)**).

Could you please clarify?

**Comment by nzakas** [comment]
No minimal reproduction link, so closing. This is required for bug reports.

**Comment by Slapbox** [comment]
@fasttime thanks for your reply and sorry for the delay - this seems to have been marked as read but I never saw it.

[Here's a repro using ESLint Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci1jb25zdDogXCJlcnJvclwiICovXG5cblxuICBsZXQgcHJlcG9zaXRpb247XG4gIGxldCBtdXRhYmxlU3RyaW5nID0gJ2FueXRoaW5nJztcbiAgWywgcHJlcG9zaXRpb24gPSAnJywgbXV0YWJsZVN0cmluZ10gPSBtdXRhYmxlU3RyaW5nLm1hdGNoKC8oYSkoYikoYykvKTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

@fasttime I'm afraid I don't understand your explanation - I think because my example made bad assumptions. In our real case, the right-hand side is a RegExp match.

I don't think ESLint can know that the variable is never reassigned in my updated example - but it still complains that I should turn `let` into `const`.

**Comment by fasttime** [comment]
@Slapbox I'm sorry that my explanation was not clear enough. Be assured that this is not a bug: it's intended behavior that default value syntax is not considered as a second assignment.

You are welcome to use the [discussions page](https://github.com/eslint/eslint/discussions) or our [Discord server](https://eslint.org/chat) if you have any more questions.

**Comment by DaSchTour** [comment]
I have the same issue with this small function

```
export function removeZeros(str: string, seperator: '.' | ',' = '.') {
  if (!str.includes(seperator)) {
    return str;
  }

  let [integer, decimal] = str.split(seperator);

  while (decimal.endsWith('0')) {
    decimal = decimal.substring(0, decimal.length - 1);
  }

  return decimal.length > 0 ? integer + seperator + decimal : integer;
}
```

As integer is not reassigned the linter complains but when it is changed to const then it will not work as decimal is reassigned.

**Comment by Rec0iL99** [comment]
@DaSchTour you could do something like this:

```js
export function removeZeros(str, seperator = '.') {
    if (!str.includes(seperator)) {
      return str;
    }
  
    const split = str.split(seperator);
    const [integer] = split;
    let [, decimal] = split;
  
    while (decimal.endsWith('0')) {
      decimal = decimal.substring(0, decimal.length - 1);
    }
  
    return decimal.length > 0 ? integer + seperator + decimal : integer;
}
```

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci1jb25zdDogXCJlcnJvclwiICovXG5cbmV4cG9ydCBmdW5jdGlvbiByZW1vdmVaZXJvcyhzdHIsIHNlcGVyYXRvciA9ICcuJykge1xuICAgIGlmICghc3RyLmluY2x1ZGVzKHNlcGVyYXRvcikpIHtcbiAgICAgIHJldHVybiBzdHI7XG4gICAgfVxuICBcbiAgICBjb25zdCBzcGxpdCA9IHN0ci5zcGxpdChzZXBlcmF0b3IpO1xuICAgIGNvbnN0IFtpbnRlZ2VyXSA9IHNwbGl0O1xuICAgIGxldCBbLCBkZWNpbWFsXSA9IHNwbGl0O1xuICBcbiAgICB3aGlsZSAoZGVjaW1hbC5lbmRzV2l0aCgnMCcpKSB7XG4gICAgICBkZWNpbWFsID0gZGVjaW1hbC5zdWJzdHJpbmcoMCwgZGVjaW1hbC5sZW5ndGggLSAxKTtcbiAgICB9XG4gIFxuICAgIHJldHVybiBkZWNpbWFsLmxlbmd0aCA+IDAgPyBpbnRlZ2VyICsgc2VwZXJhdG9yICsgZGVjaW1hbCA6IGludGVnZXI7XG59Iiwib3B0aW9ucyI6eyJlbnYiOnsiZXM2Ijp0cnVlfSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=)
